presto-oracle/src/main/java/com/facebook/presto/plugin/oracle/OracleClient.java

```diff
@@ -286,10 +300,56 @@ public Optional<ReadMapping> toPrestoType(ConnectorSession session, JdbcTypeHand
                 return Optional.of(varcharReadMapping(createVarcharType(columnSize)));
             case Types.VARCHAR:
                 return Optional.of(varcharReadMapping(createVarcharType(columnSize)));
+            case Types.DATE:
+            case Types.TIMESTAMP:
+                return Optional.of(timestampReadMapping());
         }
         return super.toPrestoType(session, typeHandle);
     }
 
+    @Override
+    protected String toSqlType(Type type)
+    {
+        if (isVarcharType(type)) {
+            VarcharType varcharType = (VarcharType) type;
+            if (varcharType.isUnbounded()) {
+                return "VARCHAR2(4000)"; // Oracle max before CLOB
+            }
+            return format("VARCHAR2(%s)", varcharType.getLengthSafe());
+        }
```

Comment:
Also VARCHAR2 has a hard limit of 4000 bytes, VARCHAR2 cannot safely hold more than ~1000 characters. So there should be condition here -

```
if (isVarcharType(type)) {
        VarcharType varcharType = (VarcharType) type;
        int length = varcharType.getBoundedLength();

        // Safe limit: 1000 CHAR (4000 bytes max in AL32UTF8)
        if (varcharType.isUnbounded() || length > 1000) {
            return "NCLOB";
        }
        return format("VARCHAR2(%s CHAR)", length);
}
```